### ISSUE TITLE
Class_ratio and Train_class_count

### ISSUE THREAD

**Issue by Young-eng** [Hope1337/YOWOv3#15]
Thanks for your wonderful job! I wonder how you get the values of class ratio and train class count in ava_config, 

it seems that class_ratio is not the proportion of specific label category.  For example: 79: 0.5059319458527529, while the instances of class 79 in train is 167115, is not almost a half of all instances.

**Comment by Hope1337** [comment]
Hi @Young-eng, yhank you for asking.

- Regarding the train class count:
During the experiment, I noticed that the AP score is consistently heavily skewed between classes. Therefore, I created a small script to count (which was deleted after completion). For each image in the training set, whenever a class appears, the class count for that class is incremented by one. As you can see, this dataset is severely imbalanced.

- Regarding class ratio:
With the observation above, I utilized a loss balance term (as mentioned in the paper). However, please note that the class ratio will be flipped as in line 381 [here](https://github.com/Hope1337/YOWOv3/blob/480bf188b4bd6b605a8538cfbbbdaaa11fc3be2a/utils/loss.py#L381) : 

```python
if self.mode == 'unbalance':
   ratio_dict = config['class_ratio']
   self.class_weights = torch.zeros(len(list(ratio_dict.keys())))
   for x in ratio_dict.keys():
       self.class_weights[x] = 1.0 - ratio_dict[x]

   self.class_weights = self.class_weights.to('cuda')
```

As in YOWOv3 paper, class balance term = $exp(class\\_ratio)$

We set this value such that for classes appearing more frequently, it will be closer to $0.5$, while for less common classes, it will closer to $0$ will approximately fall within the range of $(0...0.5)$, with denser classes tending towards $0.5$ and less frequent classes tending towards $0$. In this way, $exp(class\\_ratio)$ in for classes with a $class\\_ratio$ close to $0.5$ (meaning they appear more frequently), the class balance term $exp(class\\_ratio)$ will remain almost unchanged, meaning that the loss will not change significantly if mispredicted. Conversely, for classes with a class_ratio close to $0$ (indicating they are less common), the class balance term $exp(class\\_ratio)$ will heavily penalize if a class present in the ground truth is not predicted by the model, while the penalty will be much lighter if a class not in the ground truth is predicted. This creates a bias that helps improve the prediction of less common classes. 

Instead of manually assigning (or you can also write code to automatically assign) the $class_ratio$, I leverage what is predefined from [the research of the YOWO model](https://github.com/wei-tim/YOWO/blob/master/cfg/ava_categories_ratio.json) for simplicity.

**Comment by Young-eng** [comment]
Thanks for your answers!

 I wonder how I can calculate the *class_ratio*, I see that every class label has a number, like:
  79: 0.5059319458527529
  11: 0.5005430972490094
  78: 0.3317767554377424
  73: 0.32139321745252786
  10: 0.3018567607333167
  16: 0.24206491290886498
  13: 0.12267378359756043
Do you know how to calculate and get these numbers ? I am creating my custom AVA dataset. It is also unbalanced. Thanks

**Comment by Hope1337** [comment]
- You can rely on my previous answer to generate code that calculates those numbers.
- AVAv2.2 has 80 classes, but the authors only evaluated on a subset of 60 classes as mentioned in the AVA paper.
